docs/design/2021-02-09-ttl-table.md

```diff
@@ -0,0 +1,45 @@
+# Proposal: Time to live (TTL) Table
+
+- Author(s):     [Xiaoguang Sun](https://github.com/sunxiaoguang)
+- Last updated:  2021-02-09
+- Discussion at: https://github.com/pingcap/tidb/issues/22762
+
+## Abstract
+TTL Table makes it possible to automatically garbage collect data by user defined retention policy at either fine or coarse granularity.
+
+## Background
+There are application scenarios that data is only valuable for a certain period of time after ingestion and can be deleted permanently after expiration. Tracing, audit logs or push notification with expiration are examples of such applications. With the help of TTL tables, users can be relieved from taking care of the data life cycle in such tables. Therefore people are more willing to use TiDB as a general storage for such scenarios.
+
+## Proposal
+Introduce new `TTL` and `TTL_GRANULARITY` table options when creating or altering tables. By specifying `TTL`, users can expect expired data to be removed automatically. Additionally, users can also choose from `ROW` or `PARTITION` for `TTL_GRANULARITY` option to trade off collect granularity and cost to run such garbage collection. `ROW` mode evaluates expiry time against each row of data and reclaim space at row basis. This gives us the finest granularity and allows the most accurate expiration timing. `PARTITION` mode on the other hand partitions data according to its last update time. A background timer is going to rollover partitions and truncate all expired data within the oldest partition all at once.
+
+## Rationale
+Users can choose to store expiry time explicitly in the table schema and run delete on such tables periodically to reclaim space. This approach not only post burden on users but also introduced unnecessary load on TiDB to execute such updates and run GC at background to eventually reclaim space. The extra load is especially high when there is lots of data expiring within a short period of time.
+Another way to quickly remove expired data is managing all data in a carefully designed partition table. Bucketing data according to their update time makes it possible to quickly delete all expired records within a partition with a simple `TRUNCATE`. Compared to TTL Table's approach, these two options are error prone and suboptimal.
+
+## Compatibility and Migration Plan
+Due to the fact that TiDB by itself prohibits either converting ordinary table to partition table or between different partition table types, it is only possible to convert existing table to TTL table with `ROW` granularity. A TTL table with `PARTITION` granularity is implemented as a special partition table therefore conflicts with any other partition table type. Newly created tables can choose either `ROW` or `PARTITION` as a trade off between accuracy of reclaiming time and efficiency of garbage collecting expired data.
```

Comment:
Could you show the DDL syntax to:
- alter a normal table to a TTL table 
- alter a TTL table to a normal table
- alter the TTL of a TTL table

Besides, what will happen when the garbage collector is collecting the table during the DDL?

Reply:
Hm, we didn't consider the case that people want to change a TTL back to normal table. What I have right look strange, let's try finding out better way to do so.

- Alter a normal table to TTL tablem
ALTER TABLE tbl TTL='10m' TTL_GRANULARITY='ROW';

- Alter a TTL table to normal table
ALTER TABLE ttl_tbl TTL='NONE';

- Alter the TTL of a TTL table
ALTER TABLE ttl_tbl TTL='10h';